File: backend/app/services/classified_service.py
from ..repositories.classified_repo import FavoriteRepository, ListingRepository, MessageRepository
from ..repositories.users_repo import UserRepository
from ..utils.errors import ConflictException, ForbiddenException, NotFoundException
from ..utils.helpers import parse_object_id, utc_now_iso


class ClassifiedService:
    def __init__(self):
        self.listings = ListingRepository()
        self.favorites = FavoriteRepository()
        self.messages = MessageRepository()
        self.users = UserRepository()

    async def list_listings(self, search, category, condition, min_price, max_price, location,
                            sort, direction, skip, limit) -> list[dict]:
        items = await self.listings.list_all(
            search=search, category=category, condition=condition, min_price=min_price,
            max_price=max_price, location=location, sort=sort, direction=direction,
            skip=skip, limit=limit,
        )
        return await self._enrich_sellers(items)

    async def get(self, listing_id: str) -> dict:
        oid = parse_object_id(listing_id, "listing id")
        listing = await self.listings.find_by_id(oid)
        if listing is None:
            raise NotFoundException("Listing not found", "LISTING_NOT_FOUND")
        seller = await self.users.find_by_id(parse_object_id(listing["seller_id"], "seller id"))
        if seller:
            listing["seller"] = {"id": seller["id"], "name": seller["name"], "email": seller.get("email"), "bio": seller.get("bio", "")}
        else:
            listing["seller"] = {"id": listing["seller_id"], "name": "Unknown"}
        return listing

    async def my_listings(self, seller_id: str) -> list[dict]:
        return await self.listings.list_by_seller(seller_id)

    async def create(self, seller_id: str, payload: dict) -> dict:
        doc = {
            "seller_id": seller_id,
            "status": "AVAILABLE",
            "created_at": utc_now_iso(),
            "updated_at": utc_now_iso(),
            **payload,
        }
        return await self.listings.create(doc)

    async def update(self, seller_id: str, listing_id: str, payload: dict) -> dict:
        oid = parse_object_id(listing_id, "listing id")
        listing = await self.listings.find_by_id(oid)
        if listing is None:
            raise NotFoundException("Listing not found", "LISTING_NOT_FOUND")
        if listing["seller_id"] != seller_id:
            raise ForbiddenException("You can only modify your own listings")
        return await self.listings.update(oid, {**payload, "updated_at": utc_now_iso()})

    async def delete(self, seller_id: str, listing_id: str) -> None:
        oid = parse_object_id(listing_id, "listing id")
        listing = await self.listings.find_by_id(oid)
        if listing is None:
            raise NotFoundException("Listing not found", "LISTING_NOT_FOUND")
        if listing["seller_id"] != seller_id:
            raise ForbiddenException("You can only delete your own listings")
        await self.favorites.remove_for_listing(str(oid))
        await self.listings.delete(oid)

    async def mark_sold(self, seller_id: str, listing_id: str) -> dict:
        return await self.update(seller_id, listing_id, {"status": "SOLD"})

    async def toggle_favorite(self, user_id: str, listing_id: str) -> dict:
        oid = parse_object_id(listing_id, "listing id")
        listing = await self.listings.find_by_id(oid)
        if listing is None:
            raise NotFoundException("Listing not found", "LISTING_NOT_FOUND")
        ids = await self.favorites.ids_for_user(user_id)
        if str(oid) in ids:
            await self.favorites.remove(user_id, str(oid))
            return {"favorited": False}
        await self.favorites.add(user_id, str(oid))
        return {"favorited": True}

    async def my_favorites(self, user_id: str) -> list[dict]:
        return await self.favorites.listings_for_user(user_id)

    async def contact_seller(self, sender_id: str, listing_id: str, name: str, message: str) -> dict:
        oid = parse_object_id(listing_id, "listing id")
        listing = await self.listings.find_by_id(oid)
        if listing is None:
            raise NotFoundException("Listing not found", "LISTING_NOT_FOUND")
        doc = {
            "listing_id": str(oid),
            "sender_id": sender_id,
            "recipient_id": listing["seller_id"],
            "name": name.strip()[:80],
            "message": message.strip()[:1000],
            "listing_title": listing["title"],
            "created_at": utc_now_iso(),
        }
        return await self.messages.create(doc)

    async def my_messages(self, user_id: str) -> list[dict]:
        return await self.messages.list_for_user(user_id)

    async def _enrich_sellers(self, listings: list[dict]) -> list[dict]:
        for listing in listings:
            try:
                seller = await self.users.find_by_id(parse_object_id(listing["seller_id"], "seller id"))
                listing["seller_name"] = seller["name"] if seller else "Unknown"
            except Exception:
                listing["seller_name"] = "Unknown"
        return listings

File: frontend/lib/features/experiment5_classifieds/services/classified_service.dart
import '../../../core/network/api_client.dart';
import '../models/contact_message.dart';
import '../models/favorite.dart';
import '../models/listing.dart';

class ClassifiedService {
  static Future<List<Listing>> listListings({
    String? search,
    String? category,
    String? condition,
    double? minPrice,
    double? maxPrice,
    String? location,
    String sort = 'created_at',
    String direction = 'desc',
  }) async {
    final data = await ApiClient.instance.get(
      '/api/listings',
      query: {
        if (search != null && search.trim().isNotEmpty) 'search': search.trim(),
        if (category != null && category != 'ALL') 'category': category,
        if (condition != null && condition != 'ALL') 'condition': condition,
        if (minPrice != null && minPrice >= 0) 'min_price': minPrice,
        if (maxPrice != null && maxPrice >= 0) 'max_price': maxPrice,
        if (location != null && location.trim().isNotEmpty) 'location': location.trim(),
        'sort': sort,
        'direction': direction,
      },
    );
    return (data as List)
        .map((e) => Listing.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  static Future<Listing> getListing(String id) async =>
      Listing.fromJson(await ApiClient.instance.get('/api/listings/$id') as Map<String, dynamic>);

  static Future<List<Listing>> myListings() async {
    final data = await ApiClient.instance.get('/api/listings/mine');
    return (data as List)
        .map((e) => Listing.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  static Future<Listing> createListing({
    required String title,
    required String description,
    required num price,
    required String category,
    required String condition,
    required String location,
    required List<String> images,
  }) async {
    final data = await ApiClient.instance.post('/api/listings', body: {
      'title': title,
      'description': description,
      'price': price,
      'category': category,
      'condition': condition,
      'location': location,
      'images': images,
    });
    return Listing.fromJson(data as Map<String, dynamic>);
  }

  static Future<Listing> updateListing(String id, Map<String, dynamic> updates) async {
    final data = await ApiClient.instance.put('/api/listings/$id', body: updates);
    return Listing.fromJson(data as Map<String, dynamic>);
  }

  static Future<void> deleteListing(String id) async {
    await ApiClient.instance.delete('/api/listings/$id');
  }

  static Future<Listing> markSold(String id) async {
    final data = await ApiClient.instance.put('/api/listings/$id/sold');
    return Listing.fromJson(data as Map<String, dynamic>);
  }

  static Future<void> contactSeller(
    String listingId, {
    required String name,
    required String message,
  }) async {
    await ApiClient.instance.post('/api/listings/$listingId/contact', body: {
      'name': name,
      'message': message,
    });
  }

  static Future<List<Favorite>> myFavorites() async {
    final data = await ApiClient.instance.get('/api/favorites');
    return (data as List)
        .map((e) => Favorite.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  static Future<bool> toggleFavorite(String listingId) async {
    final data =
        await ApiClient.instance.post('/api/favorites/$listingId') as Map<String, dynamic>;
    return data['favorited'] == true;
  }

  static Future<List<ContactMessage>> myMessages() async {
    final data = await ApiClient.instance.get('/api/messages');
    return (data as List)
        .map((e) => ContactMessage.fromJson(e as Map<String, dynamic>))
        .toList();
  }
}